File: web/js/renderer.js
// Renderer Class - Handles all game graphics and UI rendering
class Renderer {
    constructor(canvas) {
        this.canvas = canvas;
        this.ctx = canvas.getContext('2d');
        
        // Colors for different soil types
        this.soilColors = {
            0.0: COLORS.BLACK,      // Unusable
            0.5: COLORS.LIGHT_BROWN, // Poor soil
            1.0: COLORS.BROWN,      // Normal soil
            1.5: COLORS.DARK_GREEN  // Rich soil
        };
    }
    
    renderGarden(garden) {
        for (let y = 0; y < garden.height; y++) {
            for (let x = 0; x < garden.width; x++) {
                // Calculate screen position
                const screenX = x * GRID_SIZE;
                const screenY = y * GRID_SIZE;
                
                // Get soil quality
                const soilQuality = garden.soilQuality[y][x];
                
                // Determine soil color
                const soilColor = this.getSoilColor(soilQuality);
                
                // Draw soil tile
                this.ctx.fillStyle = soilColor;
                this.ctx.fillRect(screenX, screenY, GRID_SIZE, GRID_SIZE);
                
                // Draw grid lines
                this.ctx.strokeStyle = COLORS.GRAY;
                this.ctx.lineWidth = 1;
                this.ctx.strokeRect(screenX, screenY, GRID_SIZE, GRID_SIZE);
                
                // Draw water level indicator
                if (garden.waterLevels[y][x] > 0) {
                    const waterAlpha = Math.min(1.0, garden.waterLevels[y][x] * 0.1);
                    this.ctx.fillStyle = `rgba(173, 216, 230, ${waterAlpha})`;
                    this.ctx.fillRect(screenX, screenY, GRID_SIZE, GRID_SIZE);
                }
                
                // Draw fertilizer indicator
                if (garden.fertilizerLevels[y][x] > 0) {
                    const fertColor = garden.fertilizerLevels[y][x] >= 2 ? COLORS.GREEN : COLORS.LIGHT_GREEN;
                    const fertSize = Math.min(GRID_SIZE / 4, garden.fertilizerLevels[y][x] * 2);
                    const fertX = screenX + (GRID_SIZE - fertSize) / 2;
                    const fertY = screenY + (GRID_SIZE - fertSize) / 2;
                    
                    this.ctx.fillStyle = fertColor;
                    this.ctx.beginPath();
                    this.ctx.arc(fertX + fertSize / 2, fertY + fertSize / 2, fertSize / 2, 0, Math.PI * 2);
                    this.ctx.fill();
                }
            }
        }
    }
    
    renderPlant(plant) {
        // Get plant visual properties
        const props = plant.getVisualProperties();
        
        // Calculate screen position
        const screenX = (plant.x * GRID_SIZE) + (GRID_SIZE / 2);
        const screenY = (plant.y * GRID_SIZE) + (GRID_SIZE / 2);
        
        // Calculate plant size
        const plantSize = Math.floor(props.size * GRID_SIZE * 0.8);
        
        // Draw plant based on stage
        const stage = props.stage;
        const color = props.color;
        
        switch (stage) {
            case "seed":
                // Draw seed
                this.ctx.fillStyle = COLORS.BROWN;
                this.ctx.beginPath();
                this.ctx.arc(screenX, screenY, plantSize / 2, 0, Math.PI * 2);
                this.ctx.fill();
                break;
                
            case "sprout":
                // Draw sprout
                this.ctx.fillStyle = COLORS.LIGHT_GREEN;
                this.ctx.beginPath();
                this.ctx.arc(screenX, screenY, plantSize / 2, 0, Math.PI * 2);
                this.ctx.fill();
                
                // Draw stem
                this.ctx.strokeStyle = COLORS.GREEN;
                this.ctx.lineWidth = 2;
                this.ctx.beginPath();
                this.ctx.moveTo(screenX, screenY);
                this.ctx.lineTo(screenX, screenY - plantSize / 2);
                this.ctx.stroke();
                break;
                
            case "small_plant":
                // Draw small plant
                this.ctx.fillStyle = color;
                this.ctx.beginPath();
                this.ctx.arc(screenX, screenY, plantSize / 2, 0, Math.PI * 2);
                this.ctx.fill();
                
                // Draw leaves
                for (let i = 0; i < 3; i++) {
                    const angle = (i * 120) * Math.PI / 180;
                    const leafX = screenX + Math.cos(angle) * plantSize * 0.3;
                    const leafY = screenY + Math.sin(angle) * plantSize * 0.3;
                    
                    this.ctx.fillStyle = COLORS.GREEN;
                    this.ctx.beginPath();
                    this.ctx.arc(leafX, leafY, plantSize / 4, 0, Math.PI * 2);
                    this.ctx.fill();
                }
                break;
                
            case "mature_plant":
                // Draw mature plant
                this.ctx.fillStyle = color;
                this.ctx.beginPath();
                this.ctx.arc(screenX, screenY, plantSize / 2, 0, Math.PI * 2);
                this.ctx.fill();
                
                // Draw multiple leaves
                for (let i = 0; i < 6; i++) {
                    const angle = (i * 60) * Math.PI / 180;
                    const leafX = screenX + Math.cos(angle) * plantSize * 0.4;
                    const leafY = screenY + Math.sin(angle) * plantSize * 0.4;
                    
                    this.ctx.fillStyle = COLORS.GREEN;
                    this.ctx.beginPath();
                    this.ctx.arc(leafX, leafY, plantSize / 3, 0, Math.PI * 2);
                    this.ctx.fill();
                }
                break;
                
            case "harvestable":
                // Draw harvestable plant with shine effect
                this.ctx.fillStyle = color;
                this.ctx.beginPath();
                this.ctx.arc(screenX, screenY, plantSize / 2, 0, Math.PI * 2);
                this.ctx.fill();
                
                // Draw shine effect
                const shineSize = plantSize / 3;
                this.ctx.fillStyle = COLORS.WHITE;
                this.ctx.beginPath();
                this.ctx.arc(screenX - shineSize / 3, screenY - shineSize / 3, shineSize / 2, 0, Math.PI * 2);
                this.ctx.fill();
                break;
        }
        
        // Draw mutation indicator
        if (props.mutated) {
            this.ctx.strokeStyle = COLORS.PURPLE;
            this.ctx.lineWidth = 2;
            this.ctx.beginPath();
            this.ctx.arc(screenX, screenY, plantSize / 2 + 2, 0, Math.PI * 2);
            this.ctx.stroke();
        }
        
        // Draw water level indicator
        if (props.waterLevel < props.maxWaterLevel) {
            const waterBarWidth = GRID_SIZE - 4;
            const waterBarHeight = 4;
            const waterBarX = screenX - waterBarWidth / 2;
            const waterBarY = screenY - plantSize / 2 - 8;
            
            // Background bar
            this.ctx.fillStyle = COLORS.GRAY;
            this.ctx.fillRect(waterBarX, waterBarY, waterBarWidth, waterBarHeight);
            
            // Water level
            const waterFill = (props.waterLevel / props.maxWaterLevel) * waterBarWidth;
            this.ctx.fillStyle = COLORS.BLUE;
            this.ctx.fillRect(waterBarX, waterBarY, waterFill, waterBarHeight);
        }
    }
    
    renderPlayer(player) {
        // Draw player as a simple circle
        this.ctx.fillStyle = COLORS.BLUE;
        this.ctx.beginPath();
        this.ctx.arc(player.x, player.y, PLAYER_SIZE / 2, 0, Math.PI * 2);
        this.ctx.fill();
        
        // Draw player outline
        this.ctx.strokeStyle = COLORS.WHITE;
        this.ctx.lineWidth = 2;
        this.ctx.beginPath();
        this.ctx.arc(player.x, player.y, PLAYER_SIZE / 2, 0, Math.PI * 2);
        this.ctx.stroke();
        
        // Draw direction indicator
        if (player.dx !== 0 || player.dy !== 0) {
            const directionX = player.dx * 8;
            const directionY = player.dy * 8;
            
            this.ctx.strokeStyle = COLORS.WHITE;
            this.ctx.lineWidth = 3;
            this.ctx.beginPath();
            this.ctx.moveTo(player.x, player.y);
            this.ctx.lineTo(player.x + directionX, player.y + directionY);
            this.ctx.stroke();
        }
    }
    
    renderText(text, x, y, color, size, center = false, alpha = 1.0) {
        this.ctx.save();
        this.ctx.globalAlpha = alpha;
        this.ctx.fillStyle = color;
        this.ctx.font = `${size}px Arial`;
        this.ctx.textAlign = center ? 'center' : 'left';
        this.ctx.textBaseline = 'middle';
        
        if (center) {
            this.ctx.fillText(text, x, y);
        } else {
            this.ctx.fillText(text, x, y);
        }
        
        this.ctx.restore();
    }
    
    renderShop(shop, economy) {
        // Background
        this.ctx.fillStyle = COLORS.DARK_GREEN;
        this.ctx.fillRect(100, 100, SCREEN_WIDTH - 200, SCREEN_HEIGHT - 200);
        
        // Title
        this.renderText("SHOP", SCREEN_WIDTH / 2, 130, COLORS.WHITE, 32, true);
        
        // Money display
        const moneyText = `Money: $${economy.money}`;
        this.renderText(moneyText, SCREEN_WIDTH / 2, 170, COLORS.YELLOW, 24, true);
        
        // Categories
        const categories = shop.getAllCategories();
        let yOffset = 220;
        
        for (const category of categories) {
            // Category header
            this.renderText(category.toUpperCase(), 150, yOffset, COLORS.WHITE, 20);
            yOffset += 30;
            
            // Items in category
            const items = shop.getAvailableItems(category);
            for (const item of items) {
                const itemText = `${item.name} - ${item.cost} coins`;
                this.renderText(itemText, 170, yOffset, COLORS.LIGHT_GRAY, 18);
                yOffset += 25;
            }
            
            yOffset += 20;
        }
        
        // Close instruction
        this.renderText("Press ESC to close", SCREEN_WIDTH / 2, SCREEN_HEIGHT - 120, COLORS.WHITE, 18, true);
    }
    
    renderInventory(player) {
        // Background
        this.ctx.fillStyle = COLORS.DARK_GREEN;
        this.ctx.fillRect(100, 100, SCREEN_WIDTH - 200, SCREEN_HEIGHT - 200);
        
        // Title
        this.renderText("INVENTORY", SCREEN_WIDTH / 2, 130, COLORS.WHITE, 32, true);
        
        // Player stats
        const stats = player.getInventorySummary();
        let yOffset = 180;
        
        // Seeds
        this.renderText("SEEDS:", 150, yOffset, COLORS.WHITE, 20);
        yOffset += 25;
        for (const [seedType, amount] of Object.entries(stats.seeds)) {
            const seedText = `${seedType.charAt(0).toUpperCase() + seedType.slice(1)}: ${amount}`;
            this.renderText(seedText, 170, yOffset, COLORS.LIGHT_GRAY, 18);
            yOffset += 20;
        }
        
        yOffset += 20;
        
        // Tools
        this.renderText("TOOLS:", 150, yOffset, COLORS.WHITE, 20);
        yOffset += 25;
        for (const tool of stats.tools) {
            const toolText = tool.replace('_', ' ').replace(/\b\w/g, l => l.toUpperCase());
            const color = tool === stats.selectedTool ? COLORS.YELLOW : COLORS.LIGHT_GRAY;
            this.renderText(toolText, 170, yOffset, color, 18);
            yOffset += 20;
        }
        
        yOffset += 20;
        
        // Stats
        this.renderText(`Level: ${stats.level}`, 150, yOffset, COLORS.WHITE, 20);
        yOffset += 25;
        this.renderText(`Experience: ${stats.experience}/100`, 150, yOffset, COLORS.WHITE, 20);
        yOffset += 25;
        this.renderText(`Energy: ${stats.energy}/${stats.maxEnergy}`, 150, yOffset, COLORS.WHITE, 20);
        
        // Close instruction
        this.renderText("Press ESC to close", SCREEN_WIDTH / 2, SCREEN_HEIGHT - 120, COLORS.WHITE, 18, true);
    }
    
    renderHelp() {
        // Background
        this.ctx.fillStyle = COLORS.DARK_GREEN;
        this.ctx.fillRect(100, 100, SCREEN_WIDTH - 200, SCREEN_HEIGHT - 200);
        
        // Title
        this.renderText("HELP & CONTROLS", SCREEN_WIDTH / 2, 130, COLORS.WHITE, 32, true);
        
        // Controls
        const controls = [
            "WASD or Arrow Keys: Move player",
            "Space: Interact with plants/soil",
            "E: Open shop",
            "Q: Open inventory",
            "H: Show this help",
            "ESC: Pause game or close menus",
            "",
            "GAMEPLAY:",
            "• Plant seeds in soil tiles",
            "• Water plants regularly",
            "• Use fertilizer for better growth",
            "• Harvest mature plants for money",
            "• Buy upgrades in the shop",
            "• Expand your garden area"
        ];
        
        let yOffset = 180;
        for (const control of controls) {
            if (control === "") {
                yOffset += 20;
                continue;
            }
            this.renderText(control, 150, yOffset, COLORS.LIGHT_GRAY, 18);
            yOffset += 25;
        }
        
        // Close instruction
        this.renderText("Press ESC to close", SCREEN_WIDTH / 2, SCREEN_HEIGHT - 120, COLORS.WHITE, 18, true);
    }
    
    renderPauseScreen() {
        // Semi-transparent overlay
        this.ctx.fillStyle = 'rgba(0, 0, 0, 0.5)';
        this.ctx.fillRect(0, 0, SCREEN_WIDTH, SCREEN_HEIGHT);
        
        // Pause text
        this.renderText("PAUSED", SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2 - 50, COLORS.WHITE, 48, true);
        this.renderText("Press ESC to resume", SCREEN_WIDTH / 2, SCREEN_HEIGHT / 2 + 20, COLORS.WHITE, 24, true);
    }
    
    renderNotifications(notifications) {
        let yOffset = 200;
        for (const notification of notifications) {
            if (notification.timer > 0) {
                const alpha = Math.min(1.0, notification.timer / 3.0);
                this.renderText(
                    notification.message,
                    SCREEN_WIDTH / 2,
                    yOffset,
                    COLORS.WHITE,
                    20,
                    true,
                    alpha
                );
                yOffset += 30;
            }
        }
    }
    
    getSoilColor(quality) {
        if (quality <= 0.0) return this.soilColors[0.0];
        if (quality <= 0.5) return this.soilColors[0.5];
        if (quality <= 1.0) return this.soilColors[1.0];
        return this.soilColors[1.5];
    }
    
    clear() {
        this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
    }
}

File: game/renderer.py
"""
Renderer Class
Handles all game graphics and UI rendering
"""

import pygame
from typing import Dict, List, Optional
from constants import *

class Renderer:
    """Handles all rendering for the game"""
    
    def __init__(self, screen: pygame.Surface):
        self.screen = screen
        
        # Initialize fonts
        pygame.font.init()
        self.fonts = {
            16: pygame.font.Font(None, 16),
            18: pygame.font.Font(None, 18),
            20: pygame.font.Font(None, 20),
            24: pygame.font.Font(None, 24),
            32: pygame.font.Font(None, 32),
            48: pygame.font.Font(None, 48)
        }
        
        # Colors for different soil types
        self.soil_colors = {
            0.0: BLACK,      # Unusable
            0.5: LIGHT_BROWN, # Poor soil
            1.0: BROWN,      # Normal soil
            1.5: DARK_GREEN  # Rich soil
        }
    
    def render_garden(self, garden):
        """Render the garden grid"""
        for y in range(garden.height):
            for x in range(garden.width):
                # Calculate screen position
                screen_x = x * GRID_SIZE
                screen_y = y * GRID_SIZE
                
                # Get soil quality
                soil_quality = garden.soil_quality[y][x]
                
                # Determine soil color
                soil_color = self._get_soil_color(soil_quality)
                
                # Draw soil tile
                pygame.draw.rect(
                    self.screen,
                    soil_color,
                    (screen_x, screen_y, GRID_SIZE, GRID_SIZE)
                )
                
                # Draw grid lines
                pygame.draw.rect(
                    self.screen,
                    GRAY,
                    (screen_x, screen_y, GRID_SIZE, GRID_SIZE),
                    1
                )
                
                # Draw water level indicator
                if garden.water_levels[y][x] > 0:
                    water_alpha = min(255, garden.water_levels[y][x] * 50)
                    water_surface = pygame.Surface((GRID_SIZE, GRID_SIZE))
                    water_surface.set_alpha(water_alpha)
                    water_surface.fill(LIGHT_BLUE)
                    self.screen.blit(water_surface, (screen_x, screen_y))
                
                # Draw fertilizer indicator
                if garden.fertilizer_levels[y][x] > 0:
                    fert_color = GREEN if garden.fertilizer_levels[y][x] >= 2 else LIGHT_GREEN
                    fert_size = min(GRID_SIZE // 4, garden.fertilizer_levels[y][x] * 2)
                    fert_x = screen_x + (GRID_SIZE - fert_size) // 2
                    fert_y = screen_y + (GRID_SIZE - fert_size) // 2
                    pygame.draw.circle(
                        self.screen,
                        fert_color,
                        (fert_x + fert_size // 2, fert_y + fert_size // 2),
                        fert_size // 2
                    )
    
    def render_plant(self, plant):
        """Render a plant"""
        # Get plant visual properties
        props = plant.get_visual_properties()
        
        # Calculate screen position
        screen_x = (plant.x * GRID_SIZE) + (GRID_SIZE // 2)
        screen_y = (plant.y * GRID_SIZE) + (GRID_SIZE // 2)
        
        # Calculate plant size
        plant_size = int(props["size"] * GRID_SIZE * 0.8)
        
        # Draw plant based on stage
        stage = props["stage"]
        color = props["color"]
        
        if stage == "seed":
            # Draw seed
            pygame.draw.circle(
                self.screen,
                BROWN,
                (screen_x, screen_y),
                plant_size // 2
            )
        elif stage == "sprout":
            # Draw sprout
            pygame.draw.circle(
                self.screen,
                LIGHT_GREEN,
                (screen_x, screen_y),
                plant_size // 2
            )
            # Draw stem
            pygame.draw.line(
                self.screen,
                GREEN,
                (screen_x, screen_y),
                (screen_x, screen_y - plant_size // 2),
                2
            )
        elif stage == "small_plant":
            # Draw small plant
            pygame.draw.circle(
                self.screen,
                color,
                (screen_x, screen_y),
                plant_size // 2
            )
            # Draw leaves
            for i in range(3):
                angle = (i * 120) * 3.14159 / 180
                leaf_x = screen_x + int(plant_size * 0.3 * pygame.math.Vector2(1, 0).rotate(angle)[0])
                leaf_y = screen_y + int(plant_size * 0.3 * pygame.math.Vector2(1, 0).rotate(angle)[1])
                pygame.draw.circle(
                    self.screen,
                    GREEN,
                    (leaf_x, leaf_y),
                    plant_size // 4
                )
        elif stage == "mature_plant":
            # Draw mature plant
            pygame.draw.circle(
                self.screen,
                color,
                (screen_x, screen_y),
                plant_size // 2
            )
            # Draw multiple leaves
            for i in range(6):
                angle = (i * 60) * 3.14159 / 180
                leaf_x = screen_x + int(plant_size * 0.4 * pygame.math.Vector2(1, 0).rotate(angle)[0])
                leaf_y = screen_y + int(plant_size * 0.4 * pygame.math.Vector2(1, 0).rotate(angle)[1])
                pygame.draw.circle(
                    self.screen,
                    GREEN,
                    (leaf_x, leaf_y),
                    plant_size // 3
                )
        elif stage == "harvestable":
            # Draw harvestable plant with shine effect
            pygame.draw.circle(
                self.screen,
                color,
                (screen_x, screen_y),
                plant_size // 2
            )
            # Draw shine effect
            shine_size = plant_size // 3
            pygame.draw.circle(
                self.screen,
                WHITE,
                (screen_x - shine_size // 3, screen_y - shine_size // 3),
                shine_size // 2
            )
        
        # Draw mutation indicator
        if props["mutated"]:
            pygame.draw.circle(
                self.screen,
                PURPLE,
                (screen_x, screen_y),
                plant_size // 2 + 2,
                2
            )
        
        # Draw water level indicator
        if props["water_level"] < props["max_water_level"]:
            water_bar_width = GRID_SIZE - 4
            water_bar_height = 4
            water_bar_x = screen_x - water_bar_width // 2
            water_bar_y = screen_y - plant_size // 2 - 8
            
            # Background bar
            pygame.draw.rect(
                self.screen,
                GRAY,
                (water_bar_x, water_bar_y, water_bar_width, water_bar_height)
            )
            
            # Water level
            water_fill = (props["water_level"] / props["max_water_level"]) * water_bar_width
            pygame.draw.rect(
                self.screen,
                BLUE,
                (water_bar_x, water_bar_y, water_fill, water_bar_height)
            )
    
    def render_player(self, player):
        """Render the player character"""
        # Draw player as a simple circle
        pygame.draw.circle(
            self.screen,
            BLUE,
            (int(player.x), int(player.y)),
            PLAYER_SIZE // 2
        )
        
        # Draw player outline
        pygame.draw.circle(
            self.screen,
            WHITE,
            (int(player.x), int(player.y)),
            PLAYER_SIZE // 2,
            2
        )
        
        # Draw direction indicator
        if player.dx != 0 or player.dy != 0:
            # Calculate direction
            direction_x = player.dx * 8
            direction_y = player.dy * 8
            
            pygame.draw.line(
                self.screen,
                WHITE,
                (player.x, player.y),
                (player.x + direction_x, player.y + direction_y),
                3
            )
    
    def render_text(self, text: str, x: int, y: int, color: tuple, size: int, 
                   center: bool = False, alpha: int = 255):
        """Render text on screen"""
        if size not in self.fonts:
            size = 20  # Default size
        
        font = self.fonts[size]
        text_surface = font.render(text, True, color)
        
        if alpha < 255:
            text_surface.set_alpha(alpha)
        
        if center:
            text_rect = text_surface.get_rect(center=(x, y))
            self.screen.blit(text_surface, text_rect)
        else:
            self.screen.blit(text_surface, (x, y))
    
    def render_shop(self, shop, economy):
        """Render shop interface"""
        # Background
        pygame.draw.rect(
            self.screen,
            DARK_GREEN,
            (100, 100, SCREEN_WIDTH - 200, SCREEN_HEIGHT - 200)
        )
        
        # Title
        self.render_text("SHOP", SCREEN_WIDTH // 2, 130, WHITE, 32, center=True)
        
        # Money display
        money_text = f"Money: {economy.money}"
        self.render_text(money_text, SCREEN_WIDTH // 2, 170, YELLOW, 24, center=True)
        
        # Categories
        categories = shop.get_all_categories()
        y_offset = 220
        
        for category in categories:
            # Category header
            self.render_text(category.upper(), 150, y_offset, WHITE, 20)
            y_offset += 30
            
            # Items in category
            items = shop.get_available_items(category)
            for item in items:
                item_text = f"{item['name']} - {item['cost']} coins"
                self.render_text(item_text, 170, y_offset, LIGHT_GRAY, 18)
                y_offset += 25
            
            y_offset += 20
        
        # Close instruction
        self.render_text("Press ESC to close", SCREEN_WIDTH // 2, SCREEN_HEIGHT - 120, WHITE, 18, center=True)
    
    def render_inventory(self, player):
        """Render inventory interface"""
        # Background
        pygame.draw.rect(
            self.screen,
            DARK_GREEN,
            (100, 100, SCREEN_WIDTH - 200, SCREEN_HEIGHT - 200)
        )
        
        # Title
        self.render_text("INVENTORY", SCREEN_WIDTH // 2, 130, WHITE, 32, center=True)
        
        # Player stats
        stats = player.get_inventory_summary()
        y_offset = 180
        
        # Seeds
        self.render_text("SEEDS:", 150, y_offset, WHITE, 20)
        y_offset += 25
        for seed_type, amount in stats["seeds"].items():
            seed_text = f"{seed_type.title()}: {amount}"
            self.render_text(seed_text, 170, y_offset, LIGHT_GRAY, 18)
            y_offset += 20
        
        y_offset += 20
        
        # Tools
        self.render_text("TOOLS:", 150, y_offset, WHITE, 20)
        y_offset += 25
        for tool in stats["tools"]:
            tool_text = f"{tool.replace('_', ' ').title()}"
            color = YELLOW if tool == stats["selected_tool"] else LIGHT_GRAY
            self.render_text(tool_text, 170, y_offset, color, 18)
            y_offset += 20
        
        y_offset += 20
        
        # Stats
        self.render_text(f"Level: {stats['level']}", 150, y_offset, WHITE, 20)
        y_offset += 25
        self.render_text(f"Experience: {stats['experience']}/100", 150, y_offset, WHITE, 20)
        y_offset += 25
        self.render_text(f"Energy: {stats['energy']}/{stats['max_energy']}", 150, y_offset, WHITE, 20)
        
        # Close instruction
        self.render_text("Press ESC to close", SCREEN_WIDTH // 2, SCREEN_HEIGHT - 120, WHITE, 18, center=True)
    
    def render_help(self):
        """Render help interface"""
        # Background
        pygame.draw.rect(
            self.screen,
            DARK_GREEN,
            (100, 100, SCREEN_WIDTH - 200, SCREEN_HEIGHT - 200)
        )
        
        # Title
        self.render_text("HELP & CONTROLS", SCREEN_WIDTH // 2, 130, WHITE, 32, center=True)
        
        # Controls
        controls = [
            "WASD or Arrow Keys: Move player",
            "Space: Interact with plants/soil",
            "E: Open shop",
            "Q: Open inventory",
            "H: Show this help",
            "ESC: Pause game or close menus",
            "",
            "GAMEPLAY:",
            "• Plant seeds in soil tiles",
            "• Water plants regularly",
            "• Use fertilizer for better growth",
            "• Harvest mature plants for money",
            "• Buy upgrades in the shop",
            "• Expand your garden area"
        ]
        
        y_offset = 180
        for control in controls:
            if control == "":
                y_offset += 20
                continue
            self.render_text(control, 150, y_offset, LIGHT_GRAY, 18)
            y_offset += 25
        
        # Close instruction
        self.render_text("Press ESC to close", SCREEN_WIDTH // 2, SCREEN_HEIGHT - 120, WHITE, 18, center=True)
    
    def render_pause_screen(self):
        """Render pause screen"""
        # Semi-transparent overlay
        overlay = pygame.Surface((SCREEN_WIDTH, SCREEN_HEIGHT))
        overlay.set_alpha(128)
        overlay.fill(BLACK)
        self.screen.blit(overlay, (0, 0))
        
        # Pause text
        self.render_text("PAUSED", SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2 - 50, WHITE, 48, center=True)
        self.render_text("Press ESC to resume", SCREEN_WIDTH // 2, SCREEN_HEIGHT // 2 + 20, WHITE, 24, center=True)
    
    def _get_soil_color(self, quality: float) -> tuple:
        """Get color for soil quality"""
        if quality <= 0.0:
            return self.soil_colors[0.0]
        elif quality <= 0.5:
            return self.soil_colors[0.5]
        elif quality <= 1.0:
            return self.soil_colors[1.0]
        else:
            return self.soil_colors[1.5]
